src/mscorlib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -165,16 +226,65 @@ public static void Run(ExecutionContext executionContext, ContextCallback callba
             ecsw.Undo(currentThread);
         }
 
+        internal static void RunDefaultContext(ContextCallback callback, Object state)
+        {
+            // Local copies used and passed to minimise static inializer checks in NGen'd code
+            // and extern current native thread lookup calls
+            ExecutionContext defaultContext = Default;
+            Thread currentThread = Thread.CurrentThread;
+            // Fastest path for moving from default context to default context
+            if (ExecutionContextSwitcher.CurrentIsDefault(currentThread, defaultContext))
+            {
+                try
+                {
+                    callback(state);
+                }
+                finally
+                {
+                    // Restore via Cloned finally as no exception can be thrown from context Capture
+                    // https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/finally-optimizations.md#finally-cloning
+                    ExecutionContextSwitcher.UndoToDefault(currentThread, defaultContext);
+                }
+            }
+            else
+            {
+                // Current context wasn't default; so the current context needs to be captured and restored
+                Run(defaultContext, callback, state);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        internal static void RestoreDefault(Thread currentThread, ExecutionContext defaultContext)
+        {
+            // Passed to minimise static inializer checks in NGen'd code
+            // and extern current native thread lookup calls
+            Debug.Assert(currentThread == Thread.CurrentThread);
+            Debug.Assert(defaultContext == ExecutionContext.Default);
+
+            ExecutionContext previous = currentThread.ExecutionContext;
+            currentThread.ExecutionContext = defaultContext;
+
+            // For the purposes of dealing with context change, null counts as the default EC
+            if (previous != null)
```

Comment:
It looks like this would call `OnContextChanged` when `previous == defaultContext`. Maybe this should be `previous != defaultContext && previous != null`.